presto-iceberg/src/main/java/com/facebook/presto/iceberg/CallDistributedProcedureSplitSource.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.iceberg;
+
+import com.facebook.presto.iceberg.delete.DeleteFile;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.ConnectorSplit;
+import com.facebook.presto.spi.ConnectorSplitSource;
+import com.facebook.presto.spi.SplitWeight;
+import com.facebook.presto.spi.connector.ConnectorPartitionHandle;
+import com.google.common.collect.ImmutableList;
+import com.google.common.io.Closer;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.PartitionSpecParser;
+import org.apache.iceberg.TableScan;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+import static com.facebook.presto.hive.HiveCommonSessionProperties.getAffinitySchedulingFileSectionSize;
+import static com.facebook.presto.hive.HiveCommonSessionProperties.getNodeSelectionStrategy;
+import static com.facebook.presto.iceberg.FileFormat.fromIcebergFileFormat;
+import static com.facebook.presto.iceberg.IcebergUtil.getDataSequenceNumber;
+import static com.facebook.presto.iceberg.IcebergUtil.getPartitionKeys;
+import static com.facebook.presto.iceberg.IcebergUtil.partitionDataFromStructLike;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterators.limit;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class CallDistributedProcedureSplitSource
+        implements ConnectorSplitSource
+{
+    private CloseableIterator<FileScanTask> fileScanTaskIterator;
+    private Optional<Consumer<FileScanTask>> fileScanTaskConsumer;
+
+    private final TableScan tableScan;
+    private final Closer closer = Closer.create();
+    private final double minimumAssignedSplitWeight;
+    private final ConnectorSession session;
+
+    public CallDistributedProcedureSplitSource(
```

Comment:
I think the most straightforward thing to do is to simply have a new method in `ConnectorSplitSource` that explicitly takes in `ConnectorProcedureContext` and `ConnectorDistributedProcedureHandle`.  That way we can unify the implementation here and just use the existing `IcebergSplitSource`, and I also think it would be a lot more straightforward to understand compared to overloading the Iceberg Metadata object to also include this custom split source.